jujugui/static/gui/src/app/store/env/controller-api.js

```diff
@@ -973,6 +973,104 @@ YUI.add('juju-controller-api', function(Y) {
         request: 'Credential',
         params: {entities: entities}
       }, handler);
+    },
+
+    /**
+      Create or update a single cloud credential.
+
+      @method updateCloudCredential
+      @param {String} tag The Juju tag of the credential as a string, for
+        instance "cloudcred-google_dalek@local_google". Tags for credentials
+        are usually retrieved by calling the getTagsForCloudCredentials method.
+      @params {String} authType The authentication type, like "userpass",
+        "oauth2" or just "empty".
+      @params {Object} attrs Attributes containing credential values, as an
+        object mapping strings to strings.
+      @param {Function} callback A callable that must be called once the
+        operation is performed. It will receive an error message or null if the
+        credential creation/update succeeded.
+    */
+    updateCloudCredential: function(tag, authType, attrs, callback) {
+      // Decorate the user supplied callback.
+      const handler = function(userCallback, data) {
+        if (!userCallback) {
+          console.log(
+            'data returned by Cloud.UpdateCredentials API call:', data);
+          return;
+        }
+        if (data.error) {
```

Comment:
Ah! Sorry I misread what you suggested. Yes in this case data cannot be undefined or null. If it was, that condition would have been catched up in the stack. This callback is called with this data here because data['request-id'] matches the request id of the original call. So we already know that data is an object.